manager/integration/tests/test_ha.py

```diff
@@ -2426,3 +2426,94 @@ def test_replica_failure_during_attaching(settings_reset, client, core_api, volu
+
+    Case 2: Test volume operations when engine image DaemonSet is not fully
+    deployed
+    1. Attach vol-1 to node-1. Change the number of replicas of vol-1
+       to 2. Delete the replica on node-1
+    2. Delete the pod on node-1 of the engine image DaemonSet.
+       Or delete the engine image DaemonSet and wait for Longhorn
+       to automatically recreates it.
+    3. Wait for the engine image CR state become deploying
+    4. Verify that functions (snapshot, backup, detach) are working ok
+       for vol-1
+    5. Detach vol-1
+    6. Attach vol-1 to node-1. Verify that Longhorn cannot attach vol-1 to
+       node-1 since there is no engine image on node-1
+    7. Verify that all functions (attach to other nodes, snapshot, backup,
+       expand, detach) are working ok for vol-1
+    8. Verify that vol-2 cannot be attached to any nodes because one of
+       its replicas is sitting on the node-1 which doesn't have the
+       engine image
+
+    Case 3: Test engine upgrade when engine image DaemonSet is not fully
+    deployed
+    1. Deploy a new engine image, new-ei
+    2. Detach vol-1
+    3. Verify that you can upgrade vol-1 to new-ei
+    4. Attach vol-1 to node-2
+    5. Verify that you can live upgrade vol-1 to back to default engine image
+
+    Case 4: Test replicas scheduling when engine image DaemonSet is not fully
+    deployed
+    1. Create a new volume, vol-3, with 2 replicas
+    2. Verify that replicas of vol-3 are on node-2 and node-3
+    3. Check that all functions (attach to node other than node-1, snapshot,
+       backup, expand, detach) are working ok for vol-3
+
+    Case 5: Test auto upgrade engine feature when engine image DaemonSet is
+    not fully deployed
+    1. Detach vol-1 and vol-3
+    2. Upgrade vol-1 and vol-3 to the new-ei
+    3. Attach vol-3 to node node-2
+    4. Set `Concurrent Automatic Engine Upgrade Per Node Limit` setting to 3
+    5. In a 2-min retry, verify that Longhorn upgrades the engine image of
+       vol-1 and vol-3 back to the default version and Longhorn doesn't
+       automatically upgrade vol-2's engine image since it has 1 replica
+       sitting on node-1 which doesn't have the engine image deployed
+
+    Case 6: Test DR, restoring, expanding volumes when engine image DaemonSet
+    is not fully deployed
+    1. Create a DR volume (vol-dr) of 2 replicas.
+    2. Verify that 2 replicas are on node-2 and node-3 and the DR volume
+       is attached to either node-2 or node-3.
+       Let's say it is attached to node-x
+    3. Taint node-x with the taint `key=value:NoSchedule`
+    4. Delete the pod of engine image DeamonSet on node-x. Now, the engine
+       image is missing on node-1 and node-x
+    5. Verify that the owner ID of vol-dr temporarily moves to node-y which
+       is different than node-1 and node-x.
+    6. Verify that vol-dr is auto-attached node-y.
+    7. Restore a volume from backupstore with name vol-rs and replica count
+       is 1
```

Comment:
Do we expose the volume `ownerID` in API layer? 
Or you mean the DR volume will be detached automatically then reattached to node-y? IIRC, there is no related logic in the implementation PR...

Anyway, the expected behavior of this DR volume is unclear here.

Reply:
Yeah, we don't have ownerID in the API layer. Fixed it

Yes, the DR volume will be detached automatically then reattached to node-y since the we automatically attach standby volumes to the ownerID node: https://github.com/longhorn/longhorn-manager/blob/71adb0123b077982a7fe306ec7f986963debf238/controller/volume_controller.go#L1811